File: python/alien_invasion.py
# Note: here assumed that ONLY VALID DATA will be given as a map file.
# In order to focus on the algorithm itself, I skipped all validations of
# input files and command-line arguments.

import random 

def parse_world_map(mapfile):
    world_map = {}      # {key: city, value: dict(direction=vicinity)}
    with open(mapfile, 'r') as f:
        lines = [ line.strip() for line in f if line.strip() ]
    for line in lines:
        token = [ l.split('=') for l in line.split() ]
        key = token.pop(0).pop()
        world_map[key] = token and dict(token) or dict()
    return world_map

def dump_world_map(world_map):
    print("\n----- THE REMAINING WORLD ----------")
    if not world_map: print("All cities are destroyed.")
    for city, road in world_map.items():
        print(city, ' '.join('='.join(item) for item in road.items()))

def init_alien_invasion(num_aliens, world_map):
    city_aliens = {}     # {key: invaded city, value: set(alien_ids)}
    if not world_map: return city_aliens
    for alien_id in map(str, range(1, num_aliens+1)):
        invaded = random.choice(list(world_map))
        move_alien_into_city(alien_id, invaded, city_aliens)
    return city_aliens

def move_alien_from_city(alien_id, city, city_aliens):
    city_aliens[city].remove(alien_id)
    if not city_aliens[city]: del city_aliens[city]

def move_alien_into_city(alien_id, city, city_aliens):
    if city in city_aliens:
        city_aliens[city].add(alien_id)
    else:
        city_aliens[city] = set([alien_id])

def wander_randomly(world_map, city_aliens):
    for from_city, alien_ids in city_aliens.copy().items():
        if not world_map[from_city]: continue
        for alien_id in alien_ids.copy():
            into_city = random.choice(list(world_map[from_city].values()))
            move_alien_from_city(alien_id, from_city, city_aliens)
            move_alien_into_city(alien_id, into_city, city_aliens)

def destroy_and_kill(world_map, city_aliens):
    for city, alien_ids in city_aliens.copy().items():
        if len(alien_ids) < 2: continue
        for _, vicinity in world_map[city].items():
            for direction, c in world_map[vicinity].copy().items():
                if c == city: del world_map[vicinity][direction]
        del world_map[city]
        del city_aliens[city]
        ids = [ f'alien {i}' for i in alien_ids ]
        print(f'{city} has been destroyed by {", ".join(ids[:-1])} and {ids[-1]}!')

def run(num_aliens, FILE_MAP, MAX_MOVES_ALIEN):
    world_map = parse_world_map(FILE_MAP)
    city_aliens = init_alien_invasion(num_aliens, world_map)
    destroy_and_kill(world_map, city_aliens)
    
    moves = 0
    while True:
        if moves > MAX_MOVES_ALIEN: break
        if not world_map: break
        if not city_aliens: break
        wander_randomly(world_map, city_aliens)
        destroy_and_kill(world_map, city_aliens)
        moves += 1
    dump_world_map(world_map)


if __name__ == '__main__':
    import sys 
    FILE_MAP = '../worldmap/coin_world.txt'
    MAX_MOVES_ALIEN = 10000

    if len(sys.argv) != 2: sys.exit(f'Usage: {sys.argv[0]} <num of aliens>')
    run(int(sys.argv[1]), FILE_MAP, MAX_MOVES_ALIEN)

File: go/alien_invasion.go
/* 
Note: here assumed that ONLY VALID DATA will be given as a map file.
In order to focus on the algorithm itself, I skipped all validations of
input files and command-line arguments.
*/
package main

import (
    "fmt"
    "bufio"
    "os"
    "strings"
    "strconv"
    "time"
    "math/rand"
)

type Roads map[string]string
type WorldMap map[string]Roads
type Intruders map[int]bool
type AlienSpot map[string]Intruders

func parse_world_map(mapfile string) WorldMap {
    f, err := os.Open(mapfile)
    if err != nil { panic(err) }
    defer f.Close()

    world_map := make(WorldMap)
    scanner := bufio.NewScanner(f)
    for scanner.Scan() {
        line := strings.TrimSpace(scanner.Text())
        if len(line) == 0 { continue }
        tokens := strings.Fields(line)
        city, roads := tokens[0], tokens[1:]
        world_map[city] = make(Roads)
        for _, road := range roads {
            split := strings.Split(road, "=")
            direction, dest := split[0], split[1]
            world_map[city][direction] = dest
        }
    }
    return world_map
}

func dump_world_map(world_map WorldMap) {
    fmt.Println("\n----- THE REMAINING WORLD ----------")
    if len(world_map) == 0 { fmt.Println("All cities are destroyed.") }
    for city, roads := range world_map {
        roads_string := ""
        for direction, dest := range roads {
            roads_string += fmt.Sprintf(" %s=%s", direction, dest)
        }
        fmt.Printf("%s%s\n", city, roads_string)
    }
}

func dump_destroy_event(city string, intruders Intruders) {
    ids := make([]string, 0, len(intruders))
    for id, _ := range intruders { ids = append(ids, fmt.Sprintf("alien %d", id)) }
    destroyer := strings.Join(ids[:len(ids)-1], ", ")
    destroyer += fmt.Sprintf(" and %s", ids[len(ids)-1])
    fmt.Printf("%s has been destroyed by %s!\n", city, destroyer)
}

func init_alien_invasion(num_aliens int, world_map WorldMap) AlienSpot {
    rand.Seed(time.Now().Unix())
    alien_spot := make(AlienSpot)
    if len(world_map) == 0 { return alien_spot }
    cities := make([]string, 0, len(world_map))
    for k := range world_map { cities = append(cities, k) }
    for id := 1; id <= num_aliens; id++ {
        invaded := cities[rand.Intn(len(cities))]
        move_alien_into_city(id, invaded, alien_spot)
    }
    return alien_spot
}

func move_alien_from_city(alien_id int, city string, alien_spot AlienSpot) {
    delete(alien_spot[city], alien_id)
    if len(alien_spot[city]) == 0 { delete(alien_spot, city) }
}

func move_alien_into_city(alien_id int, city string, alien_spot AlienSpot) {
    _, ok := alien_spot[city]
    if ok {
        alien_spot[city][alien_id] = true
    } else {
        alien_spot[city] = Intruders{alien_id: true}
    }
}

func wander_randomly(world_map WorldMap, alien_spot AlienSpot) {
    for from_city, intruders := range alien_spot {
        // No wandering when a alien gets trapped
        if _, ok := world_map[from_city]; !ok { continue }

        cities := make([]string, 0, len(world_map[from_city]))
        for _, v := range world_map[from_city] { cities = append(cities, v) }
        if len(cities) == 0 { continue }
        for alien_id, _ := range intruders {
            // random selection based on roads connected to cities
            into_city := cities[rand.Intn(len(cities))]
            // pull aliens out of an existing location
            move_alien_from_city(alien_id, from_city, alien_spot)
            // place aliens into the newly selected location
            move_alien_into_city(alien_id, into_city, alien_spot)
        }
    }
}

func destroy_and_kill(world_map WorldMap, alien_spot AlienSpot) {
    for city, intruders := range alien_spot {
        // if the number of aliens at the same place is less than 2, do nothing
        if len(intruders) < 2 { continue }

        // otherwise, destroy the city and kill aliens in it
        for _, vicinity := range world_map[city] {
            // finding all related routes and destroy it
            for direction, c := range world_map[vicinity] {
                if c == city { delete(world_map[vicinity], direction) }
            }
        }
        // remove the city from world map and from the alien tracker
        delete(world_map, city)
        delete(alien_spot, city)

        // print out an event log based on the problem definition
        dump_destroy_event(city, intruders)
    }
}

func main() {
    if len(os.Args) != 2 {
        fmt.Fprintf(os.Stderr, "Usage: %s <num of aliens>\n", os.Args[0])
        return
    }
    const FILE_MAP = "../worldmap/universe.txt"
    const MAX_MOVES_ALIENS = 10000
    num_aliens, _ := strconv.Atoi(os.Args[1])

    // parse the map file given
    world_map := parse_world_map(FILE_MAP)

    // initialize the aliens position
    alien_spot := init_alien_invasion(num_aliens, world_map)
    destroy_and_kill(world_map, alien_spot)

    moves := 0
    for {
        // exit conditions
        if moves > MAX_MOVES_ALIENS { break }
        if len(world_map) == 0 { break }
        if len(alien_spot) == 0 { break }

        // each alien wanders randomly each turn
        wander_randomly(world_map, alien_spot)

        // update world map and aliens' position by definition
        // destroy cities and kill aliens if needed.
        destroy_and_kill(world_map, alien_spot)
        moves++
    }
    dump_world_map(world_map)
}

